internal/store/generic.go

```diff
@@ -0,0 +1,405 @@
+package store
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"errors"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/flterrors"
+	"github.com/flightctl/flightctl/internal/store/model"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type IntegrationTestCallback func()
+
+// GenericStore provides generic CRUD operations for resources
+// P is a pointer to a model, for example: *model.Device
+// M is the model, for example: model.Device
+// A is the API resource, for example: api.Device
+// AL is the API list, for example: api.DeviceList
+type GenericStore[
+	P interface {
+		model.ResourceInterface
+		*M
+	},
+	M any,
```

Comment:
If we want to be more strict about the types then we can define a type then can receive only model types:
```
type Model interface {
   model.Device|model.Fleet|model.Repository ...
}
```
Then the type of M can be `Model`